src/generic/mod.rs

```diff
@@ -0,0 +1,717 @@
+#![allow(missing_docs, clippy::undocumented_unsafe_blocks)] // TODO
+
+mod impls;
+
+#[cfg(doc)]
+use core::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed, Release, SeqCst};
+use core::{
+    cell::UnsafeCell,
+    fmt,
+    marker::PhantomData,
+    mem::{self, ManuallyDrop},
+    ptr,
+    sync::atomic::Ordering,
+};
+
+use self::private::{
+    AtomicArithmeticPriv, AtomicBitOpsPriv, AtomicMinMaxPriv, AtomicablePrimitivePriv,
+};
+use crate::utils::{assert_compare_exchange_ordering, assert_load_ordering, assert_store_ordering};
+
+#[cfg_attr(docsrs, doc(cfg(feature = "generic")))]
+pub trait AtomicablePrimitive: AtomicablePrimitivePriv {}
+
+#[cfg_attr(docsrs, doc(cfg(feature = "generic")))]
+pub trait Atomicable: Sized {
+    type Value: AtomicablePrimitive;
+
+    fn to_val(self) -> Self::Value;
+
+    /// # Panics
+    ///
+    /// If the value is not valid, this method will panic.
+    ///
+    /// This method will also panic if there is no safe way to create `Self`
+    /// from `Self::Value`.
+    fn from_val(val: Self::Value) -> Self;
+
+    /// # Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// - The value must not be created in any other way than [`Atomicable::to_val`].
+    /// - The value must not be modified by operations that are not allowed by the trait.
+    /// - If the type contains lifetimes, the value must be alive and that the lifetime is [invariant][variance].
+    /// - If the type is not `Copy`, the value must be unique.
+    ///
+    /// The implementor of this method must guarantee that conversion from a
+    /// value that meets the above requirements will not result in undefined
+    /// behavior.
+    ///
+    /// [variance]: https://doc.rust-lang.org/nomicon/subtyping.html#variance
+    #[inline]
+    unsafe fn from_val_unchecked(val: Self::Value) -> Self {
+        Self::from_val(val)
+    }
+
+    // Ideally, this should be a constant, but it is a function because it must
+    // be marked as unsafe.
+    //
+    // Note: conversions that may cause undefined behavior require unsafe code,
+    // so at first glance it seems that the constant is fine, but users can
+    // actually override them by combining them with existing Atomicable
+    // implementations.
+    /// # Safety
+    ///
+    /// Calling this method is always safe.
+    ///
+    /// The implementor of this method must guarantee that atomic operations in
+    /// the [`Relaxed`](Ordering::Relaxed) ordering will never cause undefined behavior.
```

Comment:
Hmm, I noticed that this approach does not work well once RFC 2316 (https://github.com/rust-lang/rust/issues/87919) is implemented.

Reply:
Ah, it seems to work well when *combined* with a token type that has an unsafe constructor.

```rs
    // still unsafe because users can create AllowRelaxed(true) by combining them with existing implementations.
    unsafe fn allow_relaxed() -> AllowRelaxed {
        AllowRelaxed::disallow()
    }
}

pub struct AllowRelaxed(bool);

impl AllowRelaxed {
    pub const unsafe fn allow() -> Self {
        AllowRelaxed(true)
    }

    pub const fn disallow() -> Self {
        AllowRelaxed(false)
    }
}
```